File: bitmovin/ios/Connector/BitmovinHandler.swift
//
//  BitmovinConnector.swift
//

import Foundation
import THEOplayerSDK
import CoreCollector
import THEOplayerCollector

class BitmovinHandler {
    private let theoplayerCollector: THEOplayerCollector.THEOplayerCollectorApi
    private weak var player: THEOplayer?
    private var sourceChangeListener: EventListener?
    private var currentSourceMetadata: [String:Any]?
    
    init(player: THEOplayer, bitmovinConfig: [String:Any], defaultMetadata: [String:Any]? = nil) {
        self.player = player
        let config: AnalyticsConfig = BitmovinAdapter.parseConfig(bitmovinConfig)
        let metadata: DefaultMetadata = BitmovinAdapter.parseDefaultMetadata(defaultMetadata)
        self.theoplayerCollector = THEOplayerCollector.THEOplayerCollectorFactory.create(config: config, defaultMetadata: metadata)
        self.attachListeners()
        log("Bitmovin Connector initialised with config: \(bitmovinConfig) and default metadata: \(defaultMetadata ?? [:])")
    }
    
    func updateSourceMetadata(_ sourceMetadata: [String:Any]) -> Void {
        self.currentSourceMetadata = sourceMetadata
        log("Updated current source metadata: \(sourceMetadata)")
    }
    
    func programChange(_ sourceMetadata: [String:Any]) -> Void {
        let newSourceMetadata = BitmovinAdapter.parseSourceMetadata(sourceMetadata: sourceMetadata)
        self.theoplayerCollector.programChange(newSourceMetadata: newSourceMetadata)
        log("Notified program change with new source metadata: \(sourceMetadata)")
    }
    
    func updateCustomMetadata(_ customData: [String:Any]) -> Void {
        self.theoplayerCollector.customData = BitmovinAdapter.mergeCustomData(existingCustomData: self.theoplayerCollector.customData, newCustomData: BitmovinAdapter.parseCustomData(customData))
        log("Updated custom data: \(customData)")
    }
    
    func sendCustomDataEvent(_ customData: [String:Any]) -> Void {
        self.theoplayerCollector.sendCustomDataEvent(with: BitmovinAdapter.parseCustomData(customData))
        log("Custom data event send with custom data: \(customData)")
    }
    
    func destroy() -> Void {
        self.detachListeners()
        self.theoplayerCollector.detach()
        log("Bitmovin Connector destroyed.")
    }
    
    // MARK: SSAI part
    
    func reportAdBreakStart(adBreakMetadata: [String:Any]) -> Void {
        let newAdBreakMetadata = BitmovinAdapter.parseSSAIAdBreakMetadata(adBreakMetadata)
        self.theoplayerCollector.ssai.adBreakStart(adBreakMetadata: newAdBreakMetadata)
        log("[SSAI] adBreakStart reported, with adBreakMetadata: \(adBreakMetadata)")
    }
    
    func reportAdStart(adMetadata: [String:Any]) -> Void {
        let newAdMetadata = BitmovinAdapter.parseSSAIAdMetadata(adMetadata)
        self.theoplayerCollector.ssai.adStart(adMetadata: newAdMetadata)
        log("[SSAI] adStart reported, with adMetadata: \(adMetadata)")
    }
    
    func reportQuartileFinished(adQuartile: String, adQuartileMetadata: [String:Any]) -> Void {
        let quartile = BitmovinAdapter.parseAdQuartile(adQuartile)
        let newAdQuartileMetadata = BitmovinAdapter.parseSSAIAdQuartileMetadata(adQuartileMetadata)
        self.theoplayerCollector.ssai.adQuartileFinished(adQuartile: quartile, adQuartileMetadata: newAdQuartileMetadata)
        log("[SSAI] adQuartileFinished (\(adQuartile)) reported, with adQuartileMetadata: \(adQuartileMetadata)")
    }
    
    func reportAdBreakEnd() -> Void {
        self.theoplayerCollector.ssai.adBreakEnd()
        log("[SSAI] adBreakEnd reported.")
    }
    
    // MARK: - event handling
    private func onSourceChange(event: SourceChangeEvent) {
        guard let player = self.player/*, let source = player.source*/ else { return }
        log("Received SOURCE_CHANGE event from THEOplayer")
        
        self.theoplayerCollector.detach()
        log("Player detached from collector.")
        
        self.theoplayerCollector.sourceMetadata = BitmovinAdapter.parseSourceMetadata(sourceMetadata: self.currentSourceMetadata/*, extractedSourceMetadata: source.metadata?.metadataKeys*/)
        self.currentSourceMetadata = nil
        log("SourceMetadata updated on collector.")
        
        self.theoplayerCollector.attach(to: player)
        log("Player attached to collector.")
    }
    
    // MARK: - attach/dettach main player Listeners
    private func attachListeners() {
        guard let player = self.player else { return }
        self.sourceChangeListener = player.addEventListener(type: PlayerEventTypes.SOURCE_CHANGE, listener: self.onSourceChange)
    }
    
    private func detachListeners() {
        guard let player = self.player else { return }
        if let sourceChangeListener = self.sourceChangeListener {
            player.removeEventListener(type: PlayerEventTypes.SOURCE_CHANGE, listener: sourceChangeListener)
        }
    }
}

File: bitmovin/android/src/main/java/com/theoplayer/reactnative/bitmovin/BitmovinHandler.kt
package com.theoplayer.reactnative.bitmovin

import android.content.Context
import android.util.Log
import com.bitmovin.analytics.api.AnalyticsConfig
import com.bitmovin.analytics.api.CustomData
import com.bitmovin.analytics.api.DefaultMetadata
import com.bitmovin.analytics.api.SourceMetadata
import com.bitmovin.analytics.theoplayer.api.ITHEOplayerCollector
import com.theoplayer.android.api.event.EventListener
import com.theoplayer.android.api.event.player.PlayerEventTypes
import com.theoplayer.android.api.event.player.SourceChangeEvent
import com.theoplayer.android.api.player.Player
import com.theoplayer.android.api.source.metadata.MetadataDescription
import com.theoplayer.reactnative.bitmovin.BitmovinAdapter.parseCustomDataFromJSON
import org.json.JSONObject
import kotlin.apply

class BitmovinHandler(
  context: Context,
  private val player: Player,
  config: AnalyticsConfig,
  defaultMetadata: DefaultMetadata?
) {

  private val collector: ITHEOplayerCollector = ITHEOplayerCollector.create(context, config)
  private val onSourceChange = EventListener<SourceChangeEvent> { handleSourceChange() }
  var currentSourceMetadata: SourceMetadata? = null

  init {
    player.addEventListener(PlayerEventTypes.SOURCECHANGE, onSourceChange)
    defaultMetadata?.let {
      collector.defaultMetadata = it
    }
  }

  var customData: CustomData
    get() = collector.customData
    set(value) {
      collector.customData = value
    }

  fun programChange(sourceMetadata: SourceMetadata) {
    collector.programChange(sourceMetadata)
  }

  fun sendCustomDataEvent(customData: CustomData) {
    collector.sendCustomDataEvent(customData)
  }

  private fun handleSourceChange() {
    Log.d("BitmovinConnector", "Handling source change event")
    // Detach player before setting new SourceMetadata.
    collector.detachPlayer()
    /**
     * Merge the current source metadata with the new metadata from the player source, if available.
     * Prioritize the player source's metadata.
     */
    mergeSourceMetadata(currentSourceMetadata, player.source?.metadata)?.let {
      collector.sourceMetadata = it
    }
    currentSourceMetadata = null
    collector.attachPlayer(player)
  }

  private fun mergeSourceMetadata(
    sourceMetadata: SourceMetadata?, metadata: MetadataDescription?
  ): SourceMetadata? {
    if (metadata == null) return sourceMetadata
    return SourceMetadata.Builder().apply {
      setTitle(metadata.get<String>("title") ?: sourceMetadata?.title)
      setVideoId(metadata.get<String>("videoId") ?: sourceMetadata?.videoId)
      setPath(metadata.get<String>("path") ?: sourceMetadata?.path)
      setCdnProvider(metadata.get<String>("cdnProvider") ?: sourceMetadata?.cdnProvider)
      setIsLive(metadata.get<Boolean>("isLive") ?: sourceMetadata?.isLive)
      val customData = (metadata.get<Any>("customData") as? JSONObject?)?.let {
        parseCustomDataFromJSON(
          json = it,
          buildUpon = sourceMetadata?.customData?.buildUpon())
      } ?: sourceMetadata?.customData
      customData?.let { setCustomData(it) }
    }.build()
  }

  fun destroy() {
    collector.detachPlayer()
    player.removeEventListener(PlayerEventTypes.SOURCECHANGE, onSourceChange)
  }
}
